src/app/services/project_service.js

```diff
@@ -716,6 +718,53 @@ angular
         return databases;
     }
 
+
+    function buildGroupTree(nodes, select) {
+        var groups = {}
+
+        _.each(nodes, function(node) {
+            var show = _.get(node, ['docs', 'show'], true);
+            var exclude_nodes = ['source', 'exposure', 'seed', 'macro']
+            if (node.resource_type in exclude_nodes || !show || node.access === "private") {
+                return;
+            }
+
+            var name = node.name;
+
+            var group = node.group;
+
+            var is_active = node.unique_id == select;
+
+            if (!groups[group]) {
+                groups[group] = {
+                    type: "group",
+                    name: group,
+                    active: is_active,
+                    items: []
+                };
+            } else if (is_active) {
+                groups[group].active = true;
+            }
+
+            groups[group].items.push({
+                type: 'file',
+                name: name,
+                node: node,
+                active: is_active,
+                unique_id: node.unique_id,
+                node_type: 'group'
+            })
+        });
+
+        var groups = _.sortBy(_.values(groups), 'name');
+
+        _.each(groups, function(groups) {
+            groups.items = _.sortBy(groups.items, 'name');
+        });
+
+        return groups
+    }
```

Comment:
Please please please test this, but this should be a slightly cleaner version of what you're looking for, except for the fact that it doesn't return `groups` as sorted, but I'm not sure what that mens for an object anyway.

```suggestion
    function buildGroupTree(nodes, select) {
        return nodes.reduce((acc, node) => {
            const {name, group, unique_id: id, resource_type: resourceType} = node;
            const show = _.get(node, ['docs', 'show'], true);
            const excludeNodes = ['source', 'exposure', 'seed', 'macro']

            if (node.resourceType in excludeNodes || !show || node.access === "private") {
                return;
            }

            const isActive = id == select;

            acc[group] = {
                type: "group",
                name: group,
                items: [],
                ...groups[group],
                active: isActive,
            };

            acc[group].items.push({
                type: 'file',
                name: name,
                node: node,
                active: isActive,
                unique_id: unique_id,
                node_type: 'group'
            });

            acc[group].items.sort();

            return acc;
        }, {});
    }
```

Reply:
Also please add any questions about the above and I'll explain them.